### DISCUSSION TITLE
Stop using xcpretty for scan junit generation

### DISCUSSION THREAD

**Discussion by yakimant** [fastlane/fastlane#17438] [Ideas]
xcprerry is not updated for several years, looks like it's loosing support:
xcpretty/xcpretty#360

It generates junit from xcodebuild log, which can cause issues.
Sometimes it misses some failed tests, which is dangerous for CI.

Where are other tools, which rely on xcresults for junit generation and more reliable (fastlane-community/trainer, TitouanVanBelle/XCTestHTMLReport).

**Comment by rogerluan** [comment]
Hey @yakimant 👋 I believe you can disable xcpretty by passing `disable_xcpretty` argument, or `DISABLE_XCPRETTY` env var - have you tried these options?

**Reply by yakimant** [reply]
Hi @rogerluan! Disabling the xcpretty is not a problem, but then junit xml will not be generated.

We need to switch to another tool, which will generate junit xml based on xcresults, not xcodebuild log.

**Reply by rogerluan** [reply]
Ooh, got it. Do you see alternatives out there that we could incorporate into _fastlane_ (to, say, replace xcpretty, for instance)?

**Reply by Cyberbeni** [reply]
For generating junit report: https://github.com/fastlane-community/trainer
For formatting xcodebuild output: https://github.com/thii/xcbeautify

**Reply by rogerluan** [reply]
Thanks @Cyberbeni ! I think that should be the accepted answer. Although it's a plugin, it does its job. Maybe we can refactor _fastlane_'s logic to incorporate that plugin into "native" _fastlane_, but that's beyond our scope for now AFAICT. Nonetheless, any PR that achieves this goal will be highly appreciated ❤️

**Reply by rogerluan** [reply]
cc @yakimant ☝️
🤗

**Reply by yakimant** [reply]
@rogerluan, I started this topic, because usage of xcpretty is dangerous.
It is outdated, not supported and has issues.

For example, it misses failed tests in junit file sometimes.

There is a workaround to disable xcpretty and use the tools, which @Cyberbeni mentioned.
It would be nice to have is as default.

Thanks!

**Reply by Cyberbeni** [reply]
@yakimant 

> Maybe we can refactor fastlane's logic to incorporate that plugin into "native" fastlane, but that's beyond our scope for now AFAICT. Nonetheless, any PR that achieves this goal will be highly appreciated

Don't worry, apart from a handful of people, nobody uses Ruby for their job and sweat blood having to google the syntax for every link they want to write when the fastest way to accomplish what they want is through contributing to fastlane.
